db/schema_applier.cc

```diff
@@ -745,92 +738,113 @@ static future<> merge_aggregates(distributed<service::storage_proxy>& proxy, con
-    auto&& new_column_families = co_await read_tables_for_keyspaces(proxy, keyspaces, table_kind::table, affected_tables);
-    auto&& new_types = co_await read_schema_for_keyspaces(proxy, TYPES, keyspaces);
-    auto&& new_views = co_await read_tables_for_keyspaces(proxy, keyspaces, table_kind::view, affected_tables);
-    auto new_functions = co_await read_schema_for_keyspaces(proxy, FUNCTIONS, keyspaces);
-    auto new_aggregates = co_await read_schema_for_keyspaces(proxy, AGGREGATES, keyspaces);
-    auto new_scylla_aggregates = co_await read_schema_for_keyspaces(proxy, SCYLLA_AGGREGATES, keyspaces);
-
-    std::set<sstring> keyspaces_to_drop = co_await merge_keyspaces(proxy, std::move(old_keyspaces), std::move(new_keyspaces));
-    auto types_to_drop = co_await merge_types(proxy, std::move(old_types), std::move(new_types));
-    co_await merge_tables_and_views(proxy, sys_ks,
-        std::move(old_column_families), std::move(new_column_families),
-        std::move(old_views), std::move(new_views), reload, has_tablet_mutations);
-    co_await merge_functions(proxy, std::move(old_functions), std::move(new_functions));
-    co_await merge_aggregates(proxy, std::move(old_aggregates), std::move(new_aggregates), std::move(old_scylla_aggregates), std::move(new_scylla_aggregates));
+future<> schema_applier::update(const std::vector<mutation>& muts) {
+    _after = co_await get_schema_complete_view();
+
+    std::set<sstring> keyspaces_to_drop = co_await merge_keyspaces(_proxy, _before.keyspaces, _after.keyspaces);
+    auto types_to_drop = co_await merge_types(_proxy, _before.types, _after.types);
+    co_await merge_tables_and_views(_proxy, _sys_ks,
+            _before.tables, _after.tables,
+            _before.views, _after.views,
+            _reload, _has_tablet_mutations);
+    co_await merge_functions(_proxy, _before.functions, _after.functions);
+    co_await merge_aggregates(_proxy, _before.aggregates, _after.aggregates,
+            _before.scylla_aggregates, _after.scylla_aggregates);
     co_await types_to_drop.drop();
 
-    auto& sharded_db = proxy.local().get_db();
+    auto& sharded_db = _proxy.local().get_db();
     // it is safe to drop a keyspace only when all nested ColumnFamilies where deleted
     for (auto keyspace_to_drop : keyspaces_to_drop) {
         co_await replica::database::drop_keyspace_on_all_shards(sharded_db, keyspace_to_drop);
     }
 }
 
+void schema_applier::commit() {
+    // TODO: add copy on write to schema changes
+}
+
+future<> schema_applier::notify() {
+    // TODO: pull out notifications code from update() and place here
+    co_return;
+}
+
+static future<> do_merge_schema(distributed<service::storage_proxy>& proxy, sharded<db::system_keyspace>& sys_ks, std::vector<mutation> mutations, bool reload)
+{
+    slogger.trace("do_merge_schema: {}", mutations);
+    schema_applier ap(proxy, sys_ks, reload);
+    co_await ap.prepare(mutations);
+    co_await ap.adjust(mutations);
```

Comment:
Every method seems to work with this `mutations`, can we pass this in the constructor and make it a member?

Also this interface seems a bit too complex for its own good. Are some future users expected to call these in different order or call only a subset of this? If not, why not just have a single external facing method?

Reply:
I should give some explanation behind the interface structure:
- the main purpose is to take this call out of the code: `co_await proxy.local().get_db().local().apply` so for compatibility with current code it has to have at least 2 functions: one called before apply and one after

- calls will be finally placed in raft code in a single place (but it's step by step change)

- mutations are not placed inside schema_applier because apply() mentioned above uses them

- what I try to demonstrate by current calls is that both prepare() and update() **don't modify** mutations vector
- while adjust() **does modify** it and should be considered a hack

I welcome suggestions if there is a better way to show/structure the above

As for order, yes all components should expect the same order, but they may have some calls empty (i.e. no prepare phase). Keep in mind this calling code will be used once in raft and other subsystems will simply implement the applier interface.